<div align="center">
    
  <b>Curso: Programación para el Análisis de Datos<span></b>
    
  <b>Ejercicio 1: Estructuras de Datos y DataFrames</b>
    
  <b>Presentado por: Juan Camilo Betancourt Monsalve</b>
    
  <b>Grupo: 30 </b>
    
  <b>Código: 203008069
    
  <b>Presentado a: Miguel Angel Vargas Valencia</b>
    
  <img src="https://estudios.unad.edu.co/images/footer/logo-unad-acreditacion-min.png" alt="logo universidad UNAD">
    
  
  <b>17/03/2024</b>

</div>

# Resumen

Se utiliza la librería de Python Pandas para manipular y transformar datos, así como estructuras de control para agrupar datos de formas convenientes.


**Palabras clave:** NumPy, Pandas, DataFrame, Serie, Panel, Carga de base de datos, Listas, tuplas,diccionarios

In [ ]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_breast_cancer

# Ejercicio 1: Estructuras de Datos y DataFrames

Se implementa diferentes funciones, paquetes y módulos para la computación matemática y ciencia de
datos.

## Teoría

Una revisión teórica de Estructuras de Datos y DataFrames.

### Estructuras de datos nativas de Python

Según [Wes] :

**Tupla**

Las tuplas son secuencias *inmutables* de longitud fija.


```
my_tuple = (e,e2, ...,e_n)
my_tuple = (e,)
```

Pueden almacenar *elementos de diferente tipo*:


```
my_tuple = (2 , 2.5, True, "juan")
```

**Lista**

Secuencias con longitud variable y contenido modificable.

```
my_list = [e,e2, ...,e_n]

```

Como son mutables admiten:


```
my_list[n] = <new_value>

```


**Diccionario**

El diccionario almacena una colección de  clave:valor.
Cada clave está asociada con un valor para ser recuperado,insertado,modificado o eliminado.


```
my_dict = {key1:value1, ... , keyn:valuen}
```

Los valores del diccionario pueden ser cualquier objeto de Python, sin embargo las claves deben ser inmutables como los tipos escalares(int,float,string) o las tuplas y a su vez los elementos de la tupla deben ser inmutables.


```
my_dict2 = {  [1,3] : 'bad key' }  # Diccionario incorrecto.
```

**Set**

Un set es una colección sin orden de elementos únicos.

```
my_set = { value2 , valuen , ... , value1}
```

Soportan operaciones matemáticas como union,intersección ,diferencia y diferencia simétrica.

Algunas operaciones de los sets:



```
a = {1,2,3}
b = {1,5}

a.union(b)                  => {1,2,3,5}
a.intersection(b)           => {1}
a.difference(b)             => {2,3}
a.symmetric_difference(b)   => {2 , 3 , 5}

```




### Estructuras de datos de Pandas.

Según [Samir]

Pandas es una librería de Python para análisis de datos. Detrás tiene NumPy y estructuras de datos para procesar de manera eficiente datos, uniones y leer de diferentes fuentes.

Las estructuras básicas de Pandas son:


*   Series
*   DataFrame
*   Panel


**Series**
Una serie es un array de una dimensión que puede almacenar elementos de diferente tipo.

Las series cuentan con un índice el cual es personalizable.

In [ ]:
pd.Series({'a':5 , 'b':100})

a      5
b    100
dtype: int64

**DataFrame**

Un DataFrame es una estructura de 2 dimensiones que puede almacenar elementos de diferente tipo en cada columna. Se puede crar a partir de arrays,listas,diccionarios y series.

In [ ]:
my_dict = {'c1':[1,2,3] , 'c2':['Pereira','Cali','Bogotá']}
pd.DataFrame(my_dict)

c1       c2
0   1  Pereira
1   2     Cali
2   3   Bogotá

**Panel**

Un Panel es una estructura de datos que maneja 3 dimensiones.



> Deprecated since version 0.20.0



## Cargar df1


Cargue en la variable df1 el conjunto de datos que se presentan
en el archivo Excel, Anexo 1 - Base de Datos Excel.xlsx, que se
encuentra en la carpeta de la Unidad 1 - Tarea 1 - Fundamentos
de programación científica con Python. Tenga en cuenta la hoja,
encabezado y celdas en las que comienzan los datos. Los índices
del dataframe deben ser los valores en primera columna de la
tabla de datos.

In [ ]:
# Cargar conjunto de datos de los servicios de ambulancias:

url = 'https://github.com/jcamilobm/programacion_analisis_datos/blob/main/tarea_1/Anexo%201%20-%20Base%20de%20Datos%20EXCEL.xlsx?raw=true'
df1 = pd.read_excel( url, sheet_name=1, index_col=0 , skiprows=3 )

In [ ]:
# verifico cómo se cargo dataframe
df1.head(3)

Empresa Turno  Día de la semana Experiencia del conductor  \
Servicio                                                             
1            TAC     D                 5              Principiante   
2            TAD     N                 4              Principiante   
3            TAB     N                 4                   Experto   

          Formación del Paramédico Nivel del Paramédico  \
Servicio                                                  
1                                4                    I   
2                                3                    B   
3                                4                    B   

          Cantidad de atenciones  Distancia recorrida (Km)  \
Servicio                                                     
1                              8                      78.4   
2                              5                      46.2   
3                              6                      62.4   

          Velocidad máxima (Km/h)  Consumo de combustible (Galones)  \
Servicio                                                              
1                              82                               3.1   
2                              89                               1.9   
3                              99                               2.5   

          Tiempo en Movimiento (h)  Tiempo motor ON (h)  
Servicio                                                 
1                              2.8                 3.08  
2                              2.1                 2.31  
3                              2.6                 4.68

## Cargar df2


Cargue en la variable df2 el conjunto de datos que se presentan
en el archivo CSV, Anexo 2 - Base de Datos CSV.csv, que se
encuentra en la carpeta de la Unidad 1 - Tarea 1 - Fundamentos
de programación científica con Python. Tenga en cuenta el tipo
de separador y los encabezados de las columnas.

In [ ]:
# Cargar conjunto de datos de la base de datos 2

url = 'https://raw.githubusercontent.com/jcamilobm/programacion_analisis_datos/main/tarea_1/Anexo%202%20-%20Base%20de%20Datos%20CSV.csv'
df2 = pd.read_csv( url , sep=",")

In [ ]:
# verifico cómo se cargo dataframe
df2.head(3)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.05     37.37                27.0       3885.0           661.0   
1    -118.30     34.26                43.0       1510.0           310.0   
2    -117.81     33.78                27.0       3589.0           507.0   

   population  households  median_income  median_house_value  
0      1537.0       606.0         6.6085            344700.0  
1       809.0       277.0         3.5990            176500.0  
2      1484.0       495.0         5.7934            270500.0

## Cargar df3

Cargue los datos Breast Cancer Wisconsin Dataset que se
encuentran en la variable load_breast_cancer del paquete
sklearn, en un dataframe de pandas asignado a la variable df3.

In [ ]:
# Cargar Breast Cancer Wisconsin Dataset
data = load_breast_cancer()
df3 = pd.DataFrame(data.data , columns = data.feature_names)
df3.head(3)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38           122.8     1001.0          0.11840   
1        20.57         17.77           132.9     1326.0          0.08474   
2        19.69         21.25           130.0     1203.0          0.10960   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33            184.6   
1                 0.05667  ...         24.99          23.41            158.8   
2                 0.05999  ...         23.57          25.53            152.5   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  

[3 rows x 30 columns]

## Contruir un Panel con los tres DataFrames

Construya un Panel con los tres (3) dataframes cargados
anteriormente: df1, df2 y df3. Para construir este panel, cree un
diccionario llamado panel que en sus keys tenga los nombres de
los conjuntos de datos y los valores sean los dataframes.

In [ ]:
panel = {
    'ambulancias': df1,
    'house':df2,
    'breast cancer wisconsin':df3
}

In [ ]:
panel['ambulancias'].head(3)

Empresa Turno  Día de la semana Experiencia del conductor  \
Servicio                                                             
1            TAC     D                 5              Principiante   
2            TAD     N                 4              Principiante   
3            TAB     N                 4                   Experto   

          Formación del Paramédico Nivel del Paramédico  \
Servicio                                                  
1                                4                    I   
2                                3                    B   
3                                4                    B   

          Cantidad de atenciones  Distancia recorrida (Km)  \
Servicio                                                     
1                              8                      78.4   
2                              5                      46.2   
3                              6                      62.4   

          Velocidad máxima (Km/h)  Consumo de combustible (Galones)  \
Servicio                                                              
1                              82                               3.1   
2                              89                               1.9   
3                              99                               2.5   

          Tiempo en Movimiento (h)  Tiempo motor ON (h)  
Servicio                                                 
1                              2.8                 3.08  
2                              2.1                 2.31  
3                              2.6                 4.68

## Extraer encabezados de columnas de cada DataFrame del Panel


Programe una función que extraiga todas las listas de los
encabezados de columnas que se encuentran en un Panel y
úsela para extraer las listas de columnas del Panel del punto
anterior. Imprima los resultados de forma que se diferencien los
encabezados entre los dataframes en el Panel.

In [ ]:
def get_encabezados(panel):
  enc = {}
  for key, df in panel.items():
    enc[key] = df.columns
  return enc

def show(encabezados_panel):
  for key,value in encabezados_panel.items():
    print("==========  "+str(key)+" =============\n")
    for i in value:
      print("- "+ str(i))
    print("\n")

In [ ]:
# valido
enc = get_encabezados(panel)
show(enc)

==========  ambulancias =============

- Empresa
- Turno
- Día de la semana
- Experiencia del conductor
- Formación del Paramédico
- Nivel del Paramédico
- Cantidad de atenciones
- Distancia recorrida (Km)
- Velocidad máxima (Km/h)
- Consumo de combustible (Galones)
- Tiempo en Movimiento (h)
- Tiempo motor ON (h)


==========  house =============

- longitude
- latitude
- housing_median_age
- total_rooms
- total_bedrooms
- population
- households
- median_income
- median_house_value


==========  breast cancer wisconsin =============

- mean radius
- mean texture
- mean perimeter
- mean area
- mean smoothness
- mean compactness
- mean concavity
- mean concave points
- mean symmetry
- mean fractal dimension
- radius error
- texture error
- perimeter error
- area error
- smoothness error
- compactness error
- concavity error
- concave points error
- symmetry error
- fractal dimension error
- worst radius
- worst texture
- worst perimeter
- worst area
- worst smoothness
- worst compactne

In [ ]:
# Se puede organizar en un dataframe
def get_encabezados_v2(panel):
  dic = {}
  for key , value in panel.items():
    dic[key] =  pd.Series(value.columns)
  return pd.DataFrame(dic)

In [ ]:
# probar
df_encabezados = get_encabezados_v2(panel)
df_encabezados

ambulancias               house  \
0                            Empresa           longitude   
1                              Turno            latitude   
2                   Día de la semana  housing_median_age   
3          Experiencia del conductor         total_rooms   
4           Formación del Paramédico      total_bedrooms   
5               Nivel del Paramédico          population   
6             Cantidad de atenciones          households   
7           Distancia recorrida (Km)       median_income   
8            Velocidad máxima (Km/h)  median_house_value   
9   Consumo de combustible (Galones)                 NaN   
10          Tiempo en Movimiento (h)                 NaN   
11               Tiempo motor ON (h)                 NaN   
12                               NaN                 NaN   
13                               NaN                 NaN   
14                               NaN                 NaN   
15                               NaN                 NaN   
16                               NaN                 NaN   
17                               NaN                 NaN   
18                               NaN                 NaN   
19                               NaN                 NaN   
20                               NaN                 NaN   
21                               NaN                 NaN   
22                               NaN                 NaN   
23                               NaN                 NaN   
24                               NaN                 NaN   
25                               NaN                 NaN   
26                               NaN                 NaN   
27                               NaN                 NaN   
28                               NaN                 NaN   
29                               NaN                 NaN   

    breast cancer wisconsin  
0               mean radius  
1              mean texture  
2            mean perimeter  
3                 mean area  
4           mean smoothness  
5          mean compactness  
6            mean concavity  
7       mean concave points  
8             mean symmetry  
9    mean fractal dimension  
10             radius error  
11            texture error  
12          perimeter error  
13               area error  
14         smoothness error  
15        compactness error  
16          concavity error  
17     concave points error  
18           symmetry error  
19  fractal dimension error  
20             worst radius  
21            worst texture  
22          worst perimeter  
23               worst area  
24         worst smoothness  
25        worst compactness  
26          worst concavity  
27     worst concave points  
28           worst symmetry  
29  worst fractal dimension

## Cantidad de elementos de cada df en Panel


Programe una función que imprima la cantidad de elementos
que tienen las bases de datos en un Panel y úsela con el Panel
del punto anterior. Imprima los resultados de forma que se
diferencien los tamaños de los diferentes dataframes en el
Panel.

In [ ]:
def get_dimension(panel):
  dim = {}
  for key, df in panel.items():
    dim[key] = df.shape
  return dim

In [ ]:
dim = get_dimension(panel)
show(dim)

==========  ambulancias =============

- 2200
- 12


==========  house =============

- 3000
- 9


==========  breast cancer wisconsin =============

- 569
- 30




## Filtrar  DataFrame

Filtre el dataframe df1 para obtener las muestras de la empresa
TAC. Determine cuantos elementos tiene dicho filtro.

In [ ]:
# Obtener del df1  muestras de empresa TAC
df1[df1['Empresa'] == 'TAC']

Empresa Turno  Día de la semana Experiencia del conductor  \
Servicio                                                             
1            TAC     D                 5              Principiante   
5            TAC     N                 7              Principiante   
10           TAC     N                 3              Principiante   
11           TAC     N                 1              Principiante   
16           TAC     D                 3                Intermedio   
...          ...   ...               ...                       ...   
2181         TAC     D                 6                Intermedio   
2185         TAC     D                 1              Principiante   
2192         TAC     D                 4                Intermedio   
2194         TAC     N                 3                Intermedio   
2195         TAC     N                 7              Principiante   

          Formación del Paramédico Nivel del Paramédico  \
Servicio                                                  
1                                4                    I   
5                                2                    I   
10                               3                    B   
11                               4                    B   
16                               4                    B   
...                            ...                  ...   
2181                             2                    B   
2185                             2                    I   
2192                             2                    I   
2194                             4                    B   
2195                             3                    I   

          Cantidad de atenciones  Distancia recorrida (Km)  \
Servicio                                                     
1                              8                      78.4   
5                              8                      75.6   
10                             6                      55.0   
11                             6                      59.8   
16                             5                      50.4   
...                          ...                       ...   
2181                           5                      50.4   
2185                           8                      84.0   
2192                           8                      78.0   
2194                           6                      59.4   
2195                           7                      66.0   

          Velocidad máxima (Km/h)  Consumo de combustible (Galones)  \
Servicio                                                              
1                              82                               3.1   
5                              87                               2.8   
10                             87                               2.3   
11                             88                               2.7   
16                             93                               1.9   
...                           ...                               ...   
2181                           90                               2.3   
2185                           88                               2.9   
2192                           97                               2.7   
2194                           93                               2.5   
2195                           88                               2.4   

          Tiempo en Movimiento (h)  Tiempo motor ON (h)  
Servicio                                                 
1                              2.8                 3.08  
5                              2.8                 3.64  
10                             2.5                 3.25  
11                             2.3                 3.45  
16                             2.4                 4.08  
...                            ...                  ...  
2181                           2.4                 2.88  
2185                           3.0                 3.60  
2192               

In [ ]:
# Cuántos registros tiene el filtro
df1[df1['Empresa'] == 'TAC'].shape[0]

596

El filtro tiene **596** elementos.

## Hallar cantidad de NaN por columna

Determine cuantos elementos nulos hay en cada una de las
columnas del dataframe df2.

In [ ]:
df2.isnull().sum() # por default va en axis= 0

longitude             0
latitude              5
housing_median_age    1
total_rooms           1
total_bedrooms        1
population            3
households            2
median_income         2
median_house_value    0
dtype: int64

# Conclusiones

*   Aunque las estructuras de datos nativas de Python como las listas,tuplas y diccionarios son útiles ; en muchos casos se requiere un manejo más eficiente de los datos por lo cual es necesario utilizar NumPy para optimizar estas tareas.

* NumPy ofrece optimización al operar en secuencias de gran volumen, sin embargo Pandas además permite una manipulación y transformación de los datos más intuitiva y flexible.

* Es posible importar archivos de manera local o con url con Pandas en formatos como .csv y tipo Excel.

* Es posible agrupar diferentes Dataframes en un Panel para unificar en una sola fuente y aplicar funciones con un ciclo.

# Bibliografía

Samir, M. (2015). Mastering Pyhton for Data Science. (pp. 7-10).

Wes Mckinney. (2022). Python for Data Analysis
Data Wrangling with pandas, NumPy & Jupyter. (pp. 47-61).